[FreeBSD] TLS is not working

I'm trying to run xrdp 8904252 on FreeBSD however TLS is not working.
I'm still clearifying the problem.

```
[20150611-11:54:17] [INFO ] A connection received from: 192.168.69.1 port 50842
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: 192.168.69.1:50842 - socket: 8
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: NULL:NULL - socket: 7
ssl_tls_print_error: SSL_ERROR_WANT_READ
ssl_tls_accept: TLS connection accepted
libxrdp_force_read: error
xrdp_process_main_loop: libxrdp_process_incomming failed
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: 192.168.69.1:50842 - socket: 8
xrdp_mcs_disconnect - socket closed
[20150611-11:54:17] [ERROR] Listening socket is in wrong state we terminate listener
[20150611-11:54:17] [INFO ] A connection received from: 192.168.69.1 port 50843
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: 192.168.69.1:50843 - socket: 8
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: NULL:NULL - socket: 7
ssl_tls_print_error: SSL_ERROR_WANT_READ
ssl_tls_accept: TLS connection accepted
libxrdp_force_read: error
xrdp_process_main_loop: libxrdp_process_incomming failed
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: 192.168.69.1:50843 - socket: 8
xrdp_mcs_disconnect - socket closed
[20150611-11:54:17] [ERROR] Listening socket is in wrong state we terminate listener
[20150611-11:54:17] [INFO ] A connection received from: 192.168.69.1 port 50844
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: 192.168.69.1:50844 - socket: 8
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: NULL:NULL - socket: 7
ssl_tls_print_error: SSL_ERROR_WANT_READ
ssl_tls_accept: TLS connection accepted
libxrdp_force_read: error
xrdp_process_main_loop: libxrdp_process_incomming failed
[20150611-11:54:17] [INFO ] An established connection closed to endpoint: 192.168.69.1:50844 - socket: 8
xrdp_mcs_disconnect - socket closed
[20150611-11:54:17] [ERROR] Listening socket is in wrong state we terminate listener
```

My xrdp.ini around TLS setting:
```
# security layer can be 'tls', 'rdp' or 'negotiate'
# for client compatible layer
security_layer=tls
# X.509 certificate and private key
# openssl req -x509 -newkey rsa:2048 -nodes -keyout key.pem -out cert.pem -days 365
certificate=/usr/local/etc/xrdp/cert.pem
key_file=/usr/local/etc/xrdp/key.pem
```

--- a/common/ssl_calls.c
+++ b/common/ssl_calls.c
@@ -562,11 +562,7 @@ ssl_tls_print_error(char *func, SSL *connection, int value)
             return 1;
 
         case SSL_ERROR_WANT_READ:
-            g_writeln("ssl_tls_print_error: SSL_ERROR_WANT_READ");
-            return 0;
-
         case SSL_ERROR_WANT_WRITE:
-            g_writeln("ssl_tls_print_error: SSL_ERROR_WANT_WRITE");
             return 0;
 
         case SSL_ERROR_SYSCALL:

--- a/common/ssl_calls.c
+++ b/common/ssl_calls.c
@@ -669,13 +669,24 @@ ssl_tls_accept(struct ssl_tls *self)
         return 1;
     }
 
-    connection_status = SSL_accept(self->ssl);
+    while(1) {
+        connection_status = SSL_accept(self->ssl);
 
-    if (connection_status <= 0)
-    {
-        if (ssl_tls_print_error("SSL_accept", self->ssl, connection_status))
+        if (connection_status <= 0)
         {
-            return 1;
+            if (ssl_tls_print_error("SSL_accept", self->ssl, connection_status))
+            {
+                return 1;
+            }
+            /**
+             * retry when SSL_get_error returns:
+             *     SSL_ERROR_WANT_READ
+             *     SSL_ERROR_WANT_WRITE
+             */
+        }
+        else
+        {
+            break;
         }
     }
 
@@ -709,6 +720,11 @@ ssl_tls_disconnect(struct ssl_tls *self)
             {
                 return 1;
             }
+            /**
+             * retry when SSL_get_error returns:
+             *     SSL_ERROR_WANT_READ
+             *     SSL_ERROR_WANT_WRITE
+             */
         }
     }
     return 0;
@@ -737,23 +753,37 @@ int APP_CC
 ssl_tls_read(struct ssl_tls *tls, char *data, int length)
 {
     int status;
+    int break_flag;
 
-    status = SSL_read(tls->ssl, data, length);
+    while(1) {
+        status = SSL_read(tls->ssl, data, length);
 
-    switch (SSL_get_error(tls->ssl, status))
-    {
-        case SSL_ERROR_NONE:
+        switch (SSL_get_error(tls->ssl, status))
+        {
+            case SSL_ERROR_NONE:
+                break_flag = 1;
+                break;
+
+            case SSL_ERROR_WANT_READ:
+            case SSL_ERROR_WANT_WRITE:
+                /**
+                 * retry when SSL_get_error returns:
+                 *     SSL_ERROR_WANT_READ
+                 *     SSL_ERROR_WANT_WRITE
+                 */
+                continue;
+
+            default:
+                ssl_tls_print_error("SSL_read", tls->ssl, status);
+                status = -1;
+                break_flag = 1;
+                break;
+        }
+
+        if (break_flag)
+        {
             break;
-
-        case SSL_ERROR_WANT_READ:
-        case SSL_ERROR_WANT_WRITE:
-            status = 0;
-            break;
-
-        default:
-            ssl_tls_print_error("SSL_read", tls->ssl, status);
-            status = -1;
-            break;
+        }
     }
 
     if (SSL_pending(tls->ssl) > 0)
@@ -769,23 +799,37 @@ int APP_CC
 ssl_tls_write(struct ssl_tls *tls, const char *data, int length)
 {
     int status;
+    int break_flag;
 
-    status = SSL_write(tls->ssl, data, length);
+    while(1) {
+        status = SSL_write(tls->ssl, data, length);
 
-    switch (SSL_get_error(tls->ssl, status))
-    {
-        case SSL_ERROR_NONE:
+        switch (SSL_get_error(tls->ssl, status))
+        {
+            case SSL_ERROR_NONE:
+                break_flag = 1;
+                break;
+
+            case SSL_ERROR_WANT_READ:
+            case SSL_ERROR_WANT_WRITE:
+                /**
+                 * retry when SSL_get_error returns:
+                 *     SSL_ERROR_WANT_READ
+                 *     SSL_ERROR_WANT_WRITE
+                 */
+                continue;
+
+            default:
+                ssl_tls_print_error("SSL_write", tls->ssl, status);
+                status = -1;
+                break_flag = 1;
+                break;
+        }
+
+        if (break_flag)
+        {
             break;
-
-        case SSL_ERROR_WANT_READ:
-        case SSL_ERROR_WANT_WRITE:
-            status = 0;
-            break;
-
-        default:
-            ssl_tls_print_error("SSL_write", tls->ssl, status);
-            status = -1;
-            break;
+        }
     }
 
     return status;

